User traits editor UI

flyinghermit opened the pull request:
Add, edit or delete user traits from Web UI.


changelog: Added ability to edit user traits from Web UI.

flyinghermit committed:
editor mockup

flyinghermit committed:
add empty traits to available traits so user can select from dropdown

flyinghermit committed:
trait name and value editor

flyinghermit committed:
transform editor data to match with onSave func

flyinghermit committed:
add new trait

flyinghermit committed:
remove trait

flyinghermit committed:
add available trait name list

flyinghermit committed:
show traits header only when trait count is more than zero

flyinghermit committed:
add duplicate trait rule, add requireAll wrapper to shared rules

flyinghermit committed:
update editor design

flyinghermit committed:
move TraitsEditor to separate file, add unit test

flyinghermit committed:
Merge branch 'master' into sshah/user-traits-editor

flyinghermit committed:
fix edit story

flyinghermit committed:
update /users API to handle allTraits

flyinghermit committed:
cleanup

flyinghermit committed:
prettier format

flyinghermit committed:
move validation rule out from TraitsEditor component

flyinghermit commented on lib/web/users.go:
```diff
@@ -169,7 +169,13 @@ func updateUser(r *http.Request, m userAPIGetter) (*ui.User, error) {
+		user.SetTraits(req.AllTraits)
+	} else {
+		updateUserTraits(req, user)
+	}
```
This PR makes use of `AllTraits` field to manage user Traits. 

I am guessing we would ideally want to either utilize `AllTraits` or `Traits` (which just selectively updates specific traits.  But tentatively tagging @marcoandredinis as Github shows you added user traits editing functionality in user API. Did I miss anything here?

marcoandredinis replied:
From what I can see, Traits contains only the default used traits (logins, aws arn roles, windows logins, db users, db names, ...).
While AllTraits contains everything (I think it was introduced as part of Access Lists).

I don't have the Access List context, but given that Lisa and Roman are already your reviewers, I think you are good.

flyinghermit commented:
How the editor UI looks: 
<img width="440" alt="Screenshot 2024-06-10 at 2 17 31 PM" src="https://github.com/gravitational/teleport/assets/10177602/b2dc19f4-2e2c-4a3f-b55c-76775bf73b95">

flyinghermit committed:
remove requiredAll rule wrapper

r0mant commented on lib/web/users.go:
```diff
@@ -169,7 +169,13 @@ func updateUser(r *http.Request, m userAPIGetter) (*ui.User, error) {
 	user.SetRoles(req.Roles)
 
-	updateUserTraits(req, user)
+	// we do not want situation where one trait set (allTraits)
+	// overrides another (traits).
```
I think the comment is slightly confusing, I would rephrase to indicate that in some cases clients of this API just send builtin traits to set and some send the entire traits set to set.

r0mant commented on lib/web/users.go:
```diff
@@ -169,7 +169,13 @@ func updateUser(r *http.Request, m userAPIGetter) (*ui.User, error) {
-	updateUserTraits(req, user)
+	// we do not want situation where one trait set (allTraits)
+	// overrides another (traits).
+	if len(req.AllTraits) > 0 {
+		user.SetTraits(req.AllTraits)
+	} else {
+		updateUserTraits(req, user)
+	}
```
Is this API covered by unit tests? Can we cover this logic as well?

flyinghermit replied:
tests added in https://github.com/gravitational/teleport/pull/42620/commits/0d6b589bcda5f4c5525ffba6964a281b30d7025b#diff-182f90cb2b8f651ffc752d871baae2b6de277eec5a40fbb4cbbc28c8648b4676

r0mant commented on lib/web/users.go:
```diff
@@ -305,9 +311,10 @@ type userTraits struct {
+	Name      string              `json:"name"`
+	Roles     []string            `json:"roles"`
+	Traits    userTraits          `json:"traits"`
+	AllTraits map[string][]string `json:"allTraits"`
```
Nit: What do you think about naming this `OverrideTraits` or `SetTraits`? To make it more clear that if it's provided, it will set all user traits?

Also, can you add godocs on these fields.

flyinghermit replied:
tl;dr: I updated `Traits` to `TraitsPreset` and added Godoc for clarity. Json tags are as is since they are also used in other parts of frontend. Lmk if you still find it confusing. 

The names `Traits` and `AllTraits` predates this PR and it isn't super straightforward to update these. I took a second tour today but there are other parts in frontend codebase (I found one discovery flow and not 100% on access list) that make use of these fields.

r0mant commented on lib/web/users.go:
```diff
@@ -305,9 +311,10 @@ type userTraits struct {
+	AllTraits map[string][]string `json:"allTraits"`
 }
 
 func (r *saveUserRequest) checkAndSetDefaults() error {
```
`Traits` and `AllTraits` are mutually exclusive it seems like, we should update checkAndSetDefaults accordingly.

flyinghermit replied:
updated checkAndSetDefaults in https://github.com/gravitational/teleport/pull/42620/commits/0d6b589bcda5f4c5525ffba6964a281b30d7025b

r0mant reviewed (commented):
Also @flyinghermit what's the behavior for SSO users? The "edit" dialog is not available to them so they can't edit the traits, correct?

flyinghermit committed:
Merge branch 'master' into sshah/user-traits-editor

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/user-traits-editor

flyinghermit committed:
rename trait vars and add api unit test:
- rename traits to traitsPreset
- add unit test for allTraits. Update existing checkAndSetDefault test to use table
- add comment to user types to differentiate traits and allTraits

flyinghermit commented:
> Also what's the behavior for SSO users? The "edit" dialog is not available to them so they can't edit the traits, correct?

Correct. The "options" menu, which wraps the edit button is not shown for the SSO users.

flyinghermit committed:
fix typo's, add js comments

r0mant commented on lib/web/users.go:
```diff
@@ -104,7 +104,16 @@ func createUser(r *http.Request, m userAPIGetter, createdBy string) (*ui.User, e
+	// AllTraits is also used to delete all user traits,
+	// we explicitly check if TraitsPreset is empty so
+	// to prevent traits deletion.
+	if req.TraitsPreset == (traitsPreset{}) {
```
That's a strange check, can you make it a pointer and then check against nil? And/or have an isEmpty method on the struct that returns true/false?

flyinghermit replied:
:D
Made the field pointer so now it can be compared with `nil` https://github.com/gravitational/teleport/pull/42620/commits/1a39fd7dee18702b11d3049dc6f72d303496be99

r0mant commented on lib/web/users.go:
```diff
@@ -169,7 +178,16 @@ func updateUser(r *http.Request, m userAPIGetter) (*ui.User, error) {
+	// AllTraits is also used to delete all user traits,
+	// we explicitly check if TraitsPreset is empty so
+	// to prevent traits deletion.
+	if req.TraitsPreset == (traitsPreset{}) {
```
Same comment as above?

r0mant commented on lib/web/users.go:
```diff
@@ -317,5 +346,8 @@ func (r *saveUserRequest) checkAndSetDefaults() error {
 	if len(r.Roles) == 0 {
 		return trace.BadParameter("missing roles")
 	}
+	if len(r.AllTraits) != 0 && r.TraitsPreset != (traitsPreset{}) {
```
Same comment as above regarding the check.

r0mant reviewed (commented):
Mostly lgtm but can we fix the strange traits preset check?

flyinghermit committed:
use pointer to traitsPreset so empty struct can be compared with nil

flyinghermit commented:
Friendly ping

flyinghermit commented on web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:
```diff
@@ -0,0 +1,271 @@
+      let traitValue: Option[] = i.option as Option[];
+      newTraits[i.index] = {
+        ...newTraits[i.index],
+        [i.labelField]: [...traitValue],
```
good catch, thanks.

flyinghermit commented on web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:
```diff
@@ -0,0 +1,271 @@
+  function handleInputChange(i: InputOption) {
+    let newTraits = [...configuredTraits];
+    if (i.labelField === 'traitValues') {
+      let traitValue: Option[] = i.option as Option[];
```
This is needed because `InputOption.option` accepts both `Option | Option[];`. The alternative would be to define seperate type for input options
```
type InputOption = {
  labelField: 'traitKey';
  option: Option;
  index: number;
};

type InputOptionArray = {
  labelField: 'traitValues';
  option: Option[];
  index: number;
};
```

and do `function handleInputChange(i: InputOption | InputOptionArray) {`. But that still needs to specify type on handlechange 
```
onChange={e => {
    handleInputChange({
    option: e as Option[],
```
Not sure which one it better typed though! Any suggestions?

kimlisa replied:
i would do exactly as you proposed, i think we are lacking typing on `FieldSelectCreatable` to specify like a generic type for options

flyinghermit commented on web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:
```diff
@@ -0,0 +1,271 @@
+  }, [allTraits]);
+
+  function handleInputChange(i: InputOption) {
+    let newTraits = [...configuredTraits];
```
I have a mixed feeling on using `const` in this case. Cause, `const` semantically means the value wont change. But here we change `newTraits` value and that is only possible because its an array. So using `let` conveys that its value changes below.

kimlisa replied:
i can see how that can be confusing 😅 , but i think this [article](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const) may help you out

tldr: 

> The const declaration creates an immutable reference to a value. It does not mean the value it holds is immutable — just that the variable identifier cannot be reassigned.

flyinghermit committed:
resovle review comments:
- remove double type declaration
- move traitsToTraitsOption to be used as default value in setConfiguredTraits
- label copy update
- udpate test

flyinghermit commented on web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.test.tsx:
```diff
@@ -0,0 +1,88 @@
+      </Validation>
+    );
+    fireEvent.click(screen.getByTitle('Remove Trait'));
+    expect(setConfiguredTraits).toHaveBeenCalledWith([]);
```
This test block is not needed now as the  test case in first render is now no longer required since the `traitsToTraitsOption` is moved to  `useDialog.ts` now in https://github.com/gravitational/teleport/pull/42620/commits/314783e29151dc5c6c3035f684928298f6bf7000#diff-3e716db9176ae7ea0cef1294cddedf5c09b3d6f68b10e27ce6ecf676b49e9fcd.

flyinghermit commented on web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:
```diff
@@ -0,0 +1,271 @@
+    if (!allTraits[trait][0]) {
+      continue;
+    }
+    if (allTraits[trait].length > 0) {
```
> also is it valid to add a trait with a empty name?

It shouldn't be a valid case but I know the `tctl` does not check this.

The `if (!allTraits[trait][0]) {` block exists because when user is created using tctl, some trait gets default `null` value and some, such as `host_user_gid` gets `""`. In the latter case, the value is essentially not a user configured value so I think it we don't need show them in traits editor as a "pre-configured" traits. 

<img width="718" alt="Screenshot 2024-06-13 at 12 25 02 PM" src="https://github.com/gravitational/teleport/assets/10177602/398d3590-8b49-4bcc-beac-978e618e93a1">


Pinging @r0mant if there is any reason to allow traits with empty value.

r0mant replied:
Can't think of any.

flyinghermit committed:
remove describe

flyinghermit committed:
use const for array, bring back commented empty string checker block

flyinghermit committed:
add maxWidth to DialogContent inner content to avoid flickering when scrollbar appears

kimlisa commented on web/packages/teleport/src/Users/UserAddEdit/UserAddEdit.tsx:
```diff
@@ -80,37 +84,46 @@ export function UserAddEdit(props: ReturnType<typeof useDialog>) {
-              }}
-              elevated={true}
-            />
+            <Box maxWidth={690}>
```
is this maxWidth required? (we already set a maxWidth on the parent)

flyinghermit replied:
This is required for inner content (added in https://github.com/gravitational/teleport/pull/42620/commits/7d4ecd8ed7807ce61c04b1ea64dafbce77eb2903). Without this, the inner content shrinks when vertical scrollbar appears, causing flickering.

kimlisa commented on web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:
```diff
@@ -32,6 +34,9 @@ export default function useUserDialog(props: Props) {
       label: r,
     }))
   );
+  const [configuredTraits, setConfiguredTraits] = useState<TraitsOption[]>(
+    traitsToTraitsOption(props.user.allTraits)
+  );
```
otherwise, `traitsToTraits` func will needlessly run every re-render
```suggestion
  const [configuredTraits, setConfiguredTraits] = useState<TraitsOption[]>(() =>
    traitsToTraitsOption(props.user.allTraits)
  );
```

kimlisa commented on web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:
```diff
@@ -42,9 +47,16 @@ export default function useUserDialog(props: Props) {
   }
 
   function onSave() {
+    let traitsToSave = {};
```
```suggestion
    const traitsToSave = {};
```

kimlisa commented on web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:
```diff
@@ -75,11 +87,14 @@ export default function useUserDialog(props: Props) {
     fetchRoles: props.fetchRoles,
+    setConfiguredTraits,
     isNew: props.isNew,
+    allTraits: props.user.allTraits,
```
does this need to exported?

kimlisa commented on web/packages/teleport/src/Users/UserAddEdit/TraitsEditor.tsx:
```diff
@@ -0,0 +1,248 @@
+}: TraitEditorProps) {
+  function handleInputChange(i: InputOption | InputOptionArray) {
+    const newTraits = [...configuredTraits];
+    if (i.labelField === 'traitValues') {
```
we should not allow adding empty space b/c i can do this (and save it)

we can display it if it came like that, but the web UI can prevent adding further ones

<img width="435" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/c6206f99-32a7-4eeb-8805-7b2103510157">

flyinghermit replied:
I couldn't guess why would user save an empty trait value :D but I can surely avoid handing them footgun - https://github.com/gravitational/teleport/pull/42620/commits/c3b97772077a0dd6aedf919cd2e77ff1a33e5508#diff-72c47d20adf27d7885795d45757b748a86f288c7e018d043f6d0c9cac1f798ca

kimlisa reviewed (approved):
after some minor comments, lgtm

kimlisa commented on web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:
```diff
@@ -91,3 +106,22 @@ export type Props = {
+export function traitsToTraitsOption(allTraits: AllUserTraits): TraitsOption[] {
+  const newTrait = [];
+  for (let trait in allTraits) {
+    if (!allTraits[trait][0]) {
```
actually, can we add a test for this function? wouldn't this throw if `allTraits[trait] = null`?
b/c i went to my staging cluster and saw this in my traits:
<img width="516" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/0aa6af1d-9916-4faf-b518-9b10ed80b428">

kimlisa replied:
for this condition `!allTraits[trait][0]` can this be possible?

```js
logins: ["", "root"]
```

if valid, this trait will get ignored?

flyinghermit replied:
Tests added in https://github.com/gravitational/teleport/pull/42620/commits/1d38fb380b1465b91a7b0a724628a2df3f2e63ea

- `logins: ["", "root"]` will be shown as is. But `logins: [""]` will be filtered out. I am still sticking with this choice as in the backend, in its current state, that's a still default value not set by the user. 
- that didn't throw error cause `makeTraits` checks and sets empty array. Though, just to be on safe side, https://github.com/gravitational/teleport/pull/42620/commits/1e80fcff976f4a4647a2eb6f1e1c8eabfd843d43 now avoids null value explicitly.

flyinghermit committed:
resolve review comments

flyinghermit committed:
filter empty element only if length equals 1. add unit test

flyinghermit committed:
handlechange: return for empty string, trim whitespace

flyinghermit committed:
explicitely handle null value

kimlisa commented on web/packages/teleport/src/Users/UserAddEdit/useDialog.tsx:
```diff
@@ -91,3 +105,27 @@ export type Props = {
+    if (!allTraits[trait]) {
+      continue;
+    }
+    if (!allTraits[trait][0]) {
+      if (allTraits[trait].length === 1) {
+        continue;
+      }
+    }
```
```suggestion
    if (allTraits[trait].length === 1 && !allTraits[trait][0]) {
        continue;
    }
```

kimlisa reviewed (approved):
LGTM, thanks for all the changes 👍

flyinghermit committed:
shorten empty string value check block

flyinghermit committed:
Merge branch 'master' into sshah/user-traits-editor
